### DISCUSSION TITLE
TransactionScope deprecated

### DISCUSSION THREAD

**Discussion by LNArineo** [microsoft/Dynamics365Commerce.Solutions#199] [Q&A]
Hello everyone,

 We updated to version 9.46 last week and since then we have a problem with the TransactionScope. Until the update to 9.46 we were able to include the TransactionScope using the original logic. Now we get the error "multiple database connections being opened within a same transaction".  Then we were told that TransactionScope is not used anymore.

Now we have the following situation:
The original logic writes records to the database. Our logic also writes records to the database. But they have to be written to the database at the **same time**, otherwise it could lead to errors in our head office when the job that retrieves the records is executed when our logic has not yet been executed. 

If you don't use TransactionScope anymore, how can you make sure that the database records are written only after **both** logics are executed?

**Comment by Pikhulya** [comment]
@LNArineo ,

> Now we get the error "multiple database connections being opened within a same transaction". 

Can you please provide an absolute minimal repro steps, basically the code we could just compile on our side and see it in action, based on the Commerce SDK sample ?

> Then we were told that TransactionScope is not used anymore.

Can you please provide more details on where you grabbed that information from exactly ?

> But they have to be written to the database at the **same time**, otherwise it could lead to errors in our head office when the job that retrieves the records is executed when our logic has not yet been executed.

Trying to understand how the TransactionScope could change anything in regards to writing into a DB at the same time, probably you meant your data to be stored in the DB in an atomic transaction?

**Reply by LNArineo** [reply]
Hello @Pikhulya,

below I am sending you the code snippet that led to the error: 
The error is thrown when calling: await this.ExecuteNextAsync<CreateOrderFromCartResponse>(request).ConfigureAwait(false);

```c#
public sealed class CreateOrderFromCartRequestHandler : SingleAsyncRequestHandler<CreateOrderFromCartRequest>
{
        protected override async Task<Response> Process(CreateOrderFromCartRequest request)
        {
            CreateOrderFromCartResponse createOrderFromCartResponse = null;

            using (DatabaseContext databaseContext = new DatabaseContext(request.RequestContext))
            {
                // use new transaction scope to make sure that DB changes after original functionality are commited together (since FO sync client could already be executed)
                using (TransactionScope transactionScope = new TransactionScope(TransactionScopeAsyncFlowOption.Enabled))
                {
                    // execute original functionality to save the order.
                    createOrderFromCartResponse = await this.ExecuteNextAsync<CreateOrderFromCartResponse>(request).ConfigureAwait(false);

                    // only try to create subscriptions for original PendingSalesOrder
                    if (createOrderFromCartResponse != null
                        && createOrderFromCartResponse.SalesOrder != null
                        && createOrderFromCartResponse.SalesOrder.ExtensibleSalesTransactionType == ExtensibleSalesTransactionType.PendingSalesOrder)
                    {
                        await this.createSubscriptionBilling(request.RequestContext, createOrderFromCartResponse.SalesOrder).ConfigureAwait(false);
                    }

                    // commit DB changes
                    transactionScope.Complete(); 
                }
            }

            Response response = createOrderFromCartResponse;
            createOrderFromCartResponse = null;

            return response;
        }
}
```
"Can you please provide more details on where you grabbed that information from exactly ?"

We found out from support that the TransactionScope is no longer used and the documentation also states that original logic should not be nested in a new TransactionScope if possible.

"Avoid wrapping the existing CRT request and response with a TransactionScope because doing so may cause a database exception due to multiple database connections being opened within a same transaction. Also, to improve performance, avoid using TransactionScope for read scenarios. Instead, specify TransactionScopeAsyncFlowOption.Enabled to allow for correctly asynchronous calls."

https://learn.microsoft.com/en-us/dynamics365/commerce/dev-itpro/commerce-runtime-extensibility#executing-the-next-crt-handler---chain-of-handlers-1

"Trying to understand how the TransactionScope could change anything in regards to writing into a DB at the same time, probably you meant your data to be stored in the DB in an atomic transaction?"

Yes exactly, I meant that the data is only written to the database after both logics have been run through.

**Comment by Pikhulya** [comment]
Closing as there is no response.
